### PR TITLE
[back] feat: add the `order_by` parameter to the ratings API

### PR DESCRIPTION
**Pull request by lk067fn**
**related to** #1160

---

### CODE REVIEW

**Review by lfaucon — COMMENTED** [review]
Looks good to me. Thanks a lot for the contribution! Waiting for another code review from @GresilleSiffle and to reach a decision on the inverted ordering question.

File: backend/tournesol/views/ratings.py [hunk 1]
```diff
@@ -100,6 +100,38 @@ def get_queryset(self):
...
+                    .annotate(min=Func("datetime_add", function="Min"))
+                    .values("min")
+                )
+                ratings = ratings.annotate(
+                    comparison_date = Subquery(max_comparison_date)
+                ).order_by("comparison_date")
+            elif order_by == "n_comparisons":
+                ratings = ratings.order_by("-n_comparisons")
+            elif order_by == "-n_comparisons":
+                ratings = ratings.order_by("n_comparisons")
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/tournesol/views/ratings.py:130]: I'm going to check. I think I based this on how Django uses the minus symbol for ordering, but I might have made a mistake.

[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/tournesol/views/ratings.py:130]: Yes I confirm I made a mistake, `comparison_date` should be an ascending ordering (the oldest first), and `-comparison_date` a descending (the most recent first).

**Comment by GresilleSiffle** [comment]
Thank you @lk067fn for your work!

I'm going to review today, and if needed directly push fixes in this branch.

File: backend/tournesol/serializers/rating.py [hunk 2]
```diff
@@ -21,10 +21,13 @@ class ContributorRatingSerializer(ModelSerializer):
     n_comparisons = SerializerMethodField(
         help_text="Number of comparisons submitted by the current user about the current video",
     )
+    comparison_date = SerializerMethodField(
+        help_text="Date of comparison(oldest / newest) to order ratings"
+    )
```
Comment (GresilleSiffle) [inline backend/tournesol/serializers/rating.py:26]: The `comparison_date` and the comment `Date of the comparison` is redundant.

We can safely delete it. Moreover, we should avoid mentions oldest / newest as the usage is the responsability of the API consumer. Maybe this field will be displayed, and not used for ordering, etc.

I'll fix it.

File: backend/tournesol/serializers/rating.py [hunk 3]
```diff
@@ -40,6 +43,21 @@ def get_n_comparisons(self, obj):
...
+        """
+        if hasattr(obj, "comparison_date"):
+            # Use annotated field if it has been defined by the queryset
+            return obj.comparison_date
+        
+        comparisons = obj.user.comparisons.filter(
+            Q(poll=self.context["poll"])
+            & (Q(entity_1=obj.entity) | Q(entity_2=obj.entity))
+        )
+        return list(comparisons.values_list("datetime_add", flat=True))
```
Comment (GresilleSiffle) [inline backend/tournesol/serializers/rating.py:59]: This should not return a list. This method is model specific, and thus will always return one item.

If we check what the API return at `localhost:8000/docs/#/users/users_me_contributor_ratings_list/`, we currently have:

```json
      "comparison_date": [
        "2022-10-17T15:43:16.201273Z"
      ]
```

We should have

```
      "comparison_date": "2022-10-17T15:43:16.201273Z"
```

I'll fix.

**Review by GresilleSiffle — CHANGES_REQUESTED** [review]

File: backend/tournesol/serializers/rating.py (hunk 3 above)
Comment (GresilleSiffle) [reply backend/tournesol/serializers/rating.py:59]: I realized there is no field `last_compared_at` in the `ContributorRating` model.

If we want the ratings to be ordered by the last compared date, we need to fetch the `datetime_add` value directly in the `tournesol.views.ratings.get_annotated_ratings` function.

File: backend/tournesol/serializers/rating.py (hunk 2 above)
Comment (GresilleSiffle) [reply backend/tournesol/serializers/rating.py]: Fixed by https://github.com/tournesol-app/tournesol/pull/1170/commits/e55a9b88fed1e41a852567b871e2d3ee22da3084.

File: backend/tournesol/serializers/rating.py (hunk 3 above)
Comment (GresilleSiffle) [reply backend/tournesol/serializers/rating.py]: Fixed by https://github.com/tournesol-app/tournesol/pull/1170/commits/e55a9b88fed1e41a852567b871e2d3ee22da3084.

File: backend/tournesol/views/ratings.py [hunk 4]
```diff
@@ -100,6 +100,38 @@ def get_queryset(self):
...
+            elif order_by == "-comparison_date":
+                max_comparison_date = (
+                    Comparison.objects.filter(user=OuterRef("user"))
+                    .filter(Q(entity_1=OuterRef("entity")) | Q(entity_2=OuterRef("entity")))
+                    .annotate(min=Func("datetime_add", function="Min"))
+                    .values("min")
+                )
+                ratings = ratings.annotate(
+                    comparison_date = Subquery(max_comparison_date)
+                ).order_by("comparison_date")
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/tournesol/views/ratings.py]: Fixed by https://github.com/tournesol-app/tournesol/pull/1170/commits/4c3477b47f48ec9ac76f9de803ee00bbb0cba456

File: backend/tournesol/serializers/rating.py (hunk 3 above)
Comment (lk067fn) [reply backend/tournesol/serializers/rating.py]: Thank you for your review! I see you are currently working on some changes, is there anything I can do?

Comment (GresilleSiffle) [reply backend/tournesol/serializers/rating.py]: My pleasure @lk067fn 

I'm currently updating the tests. I noticed all ratings have the same `last_compared_at` value, that's why they always pass :sweat_smile: I fix this I'll just specify different comparison date in the `setUp` function.

> is there anything I can do?

I think you are good with this PR : ) thank you very much. I'll merge as soon as I reviewed everything.

**Review by GresilleSiffle — APPROVED** [review]

**Review by GresilleSiffle — CHANGES_REQUESTED** [review]
The generated front end's OpenAPI schema is not yet valid.

**Review by GresilleSiffle — APPROVED** [review]

**Pull request merged**